create cross-encoder-integration-gsoc (#1155)
qdrant/landing_page: qdrant-landing/content/articles/cross-encoder-integration-gsoc.md

---
title: Qdrant Summer of Code 2024 - ONNX Cross Encoders in Python
short_description: QSoC 2024 ONNX Cross Encoders in Python
description: A summary of my work and experience at Qdrant Summer of Code 2024.
preview_dir: /articles_data/cross-encoder-integration-gsoc/preview
small_preview_image: /articles_data/cross-encoder-integration-gsoc/icon.svg
social_preview_image: /articles_data/cross-encoder-integration-gsoc/preview/social_preview.jpg
weight: 10
author: Huong (Celine) Hoang
author_link: https://www.linkedin.com/in/celine-h-hoang/
date: 2024-09-07T08:00:00+03:00
draft: false
keywords:
  - cross-encoder
  - reranking
  - fastembed
  - qsoc'24
---

## Introduction

> Hi everyone! I’m Huong (Celine) Hoang, and I’m thrilled to share my experience working at Qdrant this summer as part of their Summer of Code 2024 program. During my internship, I worked on integrating cross-encoders into the FastEmbed library for re-ranking tasks. To clarify, this involved adding a new capability to Qdrant's powerful vector search engine, enabling more context-aware search rankings, which are particularly important for applications like question-answering systems.

Reviewer: ```suggestion
Hi everyone! I’m Huong (Celine) Hoang, and I’m thrilled to share my experience working at Qdrant this summer as part of their Summer of Code 2024 program. During my internship, I worked on integrating cross-encoders into the FastEmbed library for re-ranking tasks. This enhancement widened the capabilities of the Qdrant ecosystem, enabling developers to build more context-aware search applications, such as question-answering systems, using Qdrant's suite of libraries.
```

This project was both technically challenging and rewarding, pushing me to grow my skills in handling large-scale ONNX (Open Neural Network Exchange) model integrations, tokenization, and more. Let me take you through the journey, the lessons learned, and where things are headed next.

## Project Overview

> Qdrant is well-known for its vector search capabilities, but my task was to go one step further—introducing cross-encoders for re-ranking. Traditionally, the FastEmbed library would generate embeddings, but cross-encoders don’t do that. Instead, they provide a list of scores based on how well a query matches a list of documents. This kind of re-ranking is critical when you want to refine search results and bring the most relevant answers to the top.

Reviewer: ```suggestion
Qdrant is well known for its vector search capabilities, but my task was to go one step further — introducing cross-encoders for re-ranking. Traditionally, the FastEmbed library would generate embeddings, but cross-encoders don’t do that. Instead, they provide a list of scores based on how well a query matches a list of documents. This kind of re-ranking is critical when you want to refine search results and bring the most relevant answers to the top.
```

The project revolved around creating a new input-output scheme: text data to scores. For this, I designed a family of classes to support ONNX models. Some of the key models I worked with included Xenova/ms-marco-MiniLM-L-6-v2, Xenova/ms-marco-MiniLM-L-12-v2, and BAAI/bge-reranker, all designed for re-ranking tasks.

> An important point to mention is that FastEmbed is a minimalistic library—it doesn’t have heavy dependencies like PyTorch or TensorFlow, and as a result, it is lightweight, occupying far less storage space.

Reviewer: ```suggestion
An important point to mention is that FastEmbed is a minimalistic library: it doesn’t have heavy dependencies like PyTorch or TensorFlow, and as a result, it is lightweight, occupying far less storage space.
```

Below is a diagram that represents the overall workflow for this project, detailing the key steps from user interaction to the final output validation:

<...>

## Technical Challenges

### 1. Building a New Input-Output Scheme

> FastEmbed already had support for embeddings, but re-ranking with cross-encoders meant building a completely new family of classes. These models accept a query and a set of documents, then return a list of relevance scores. For that, I created the base classes like TextCrossEncoderBase and OnnxCrossEncoder, taking inspiration from existing text embedding models.

Reviewer: ```suggestion
FastEmbed already had support for embeddings, but re-ranking with cross-encoders meant building a completely new family of classes. These models accept a query and a set of documents, then return a list of relevance scores. For that, I created the base classes like `TextCrossEncoderBase` and `OnnxCrossEncoder`, taking inspiration from existing text embedding models.
```

One thing I had to ensure was that the new class hierarchy was user-friendly. Users should be able to work with cross-encoders without needing to know the complexities of the underlying models. For instance, they should be able to just write:

```python
from fastembed import TextCrossEncoder

encoder = TextCrossEncoder(model_name='Xenova/ms-marco-MiniLM-L-6-v2')
scores = encoder.rerank(query, documents)
```

Meanwhile, behind the scenes, we manage all the model loading, tokenization, and scoring.

### 2. Handling Tokenization for Cross-Encoders

Cross-encoders require careful tokenization because they need to distinguish between the query and the documents. This is done using token type IDs, which help the model differentiate between the two. To implement this, I configured the tokenizer to handle pairs of inputs—concatenating the query with each document and assigning token types accordingly.

> Efficient tokenization is critical to ensure the performance of the models, and I optimized it specifically for ONNX models. The lightweight nature of FastEmbed means that it runs efficiently even without requiring heavy libraries like PyTorch or TensorFlow.

Reviewer: Has already been mentioned

```suggestion
Efficient tokenization is critical to ensure the performance of the models, and I optimized it specifically for ONNX models.
```

### 5. Testing and Validation
To ensure the accuracy and performance of the models, I conducted extensive testing. I compared the output of ONNX models with their PyTorch counterparts, ensuring the conversion to ONNX was correct. A key part of this process was rigorous testing to verify the outputs and identify potential issues, such as incorrect conversions or bugs in our implementation.

For instance, a test to validate the model's output was structured as follows:
```python
def test_rerank():
    is_ci = os.getenv("CI")

    for model_desc in TextCrossEncoder.list_supported_models():
        if not is_ci and model_desc["size_in_GB"] > 1:
            continue

        model_name = model_desc["model"]
        model = TextCrossEncoder(model_name=model_name)

        query = "What is the capital of France?"
        documents = ["Paris is the capital of France.", "Berlin is the capital of Germany."]
        scores = np.array(model.rerank(query, documents))

        canonical_scores = CANONICAL_SCORE_VALUES[model_name]
        assert np.allclose(
            scores, canonical_scores, atol=1e-3
        ), f"Model: {model_name}, Scores: {scores}, Expected: {canonical_scores}"
```

> This rigorous testing process helped ensure that the ONNX models behaved as expected, comparable to models implemented using PyTorch.

Reviewer: ```suggestion
The `CANONICAL_SCORE_VALUES` were retrieved directly from the result of applying the original PyTorch models to the same input
```
